### ISSUE TITLE
DOC: No `source` link for some functions in the API reference guide

### ISSUE THREAD

**Issue by Mukulikaa** [numpy/numpy#19520]
Most functions have a `source` link along with their documentation that leads to the source of the ~~docstrings~~ functions in GitHub. For example, [numpy.ndarray](https://numpy.org/devdocs/reference/generated/numpy.ndarray.html#numpy-ndarray), [numpy.argmax](https://numpy.org/devdocs/reference/generated/numpy.argmax.html#numpy-argmax) etc. It is a feature of the Sphinx theme that NumPy uses. It makes it very easy for contributors to locate the docstrings/code and modify them.

However, some functions like [numpy.ndarray.transpose](https://numpy.org/devdocs/reference/generated/numpy.ndarray.transpose.html#numpy.ndarray.transpose), [numpy.array](https://numpy.org/devdocs/reference/generated/numpy.array.html?highlight=numpy%20array#numpy.array) etc. don't have the `source` link. These are functions whose docstrings reside separately in `numpy/core/_add_newdocs.py` because they are defined in C-extension modules. It would be nice if we could find a way to make the feature work for these ~~docstrings~~ functions as well. 
Edit: The `source` link would lead to the C file like in [numpy.ndarray](https://numpy.org/devdocs/reference/generated/numpy.ndarray.html#numpy-ndarray).

Looking at the code [here](https://github.com/numpy/numpy/blob/1cc3f07ad3e0987da7d399dd8febf01ef3b48b45/doc/source/conf.py#L367) could be a starting point.

**Comment by mattip** [comment]
Note that at least in the case of [ndarray](), the "[source]" link does not actually point to the docstring, rather to a C file.

**Comment by Mukulikaa** [comment]
Yes, I think there was extra logic added to `conf.py` to handle some of these C-extension module functions. The main purpose of the `source` link I think is to lead to the source code itself instead of the docstrings as I said. For the python source codes, the docstrings happen to reside in the functions themselves so it seems to do both.
